File: charuconstruction/charuco.py
import json
from pathlib import Path
import random

import cv2
import numpy as np


from .camera import Camera
from .media_reader import Frame
from .math import TranslationVector, RotationMatrix
from .resolution import Resolution


class Charuco:
    def __init__(
        self,
        vertical_squares_count: int,
        horizontal_squares_count: int,
        square_len: float,
        marker_len: float,
        resolution: Resolution | str,
        aruco_dict: int = cv2.aruco.DICT_7X7_1000,
        seed: int = None,
    ):
        """
        Initialize a ChArUco board with the given parameters.

        Parameters:
            vertical_squares_count (int): Number of squares in the vertical direction.
            horizontal_squares_count (int): Number of squares in the horizontal direction.
            square_len (float): Length of the squares in meters.
            marker_len (float): Length of the ArUco markers in meters (cannot be greater than square_len).
            resolution (Resolution | str): The number of pixels in one inch on the printed page.
                If a string is provided, it should be the name of a Resolution enum member (e.g., "DPI_100").
            aruco_dict (int): Predefined ArUco dictionary to use.
            seed (int): Seed for random number generator (will determine the order of aruco markers).
        """

        if marker_len > square_len:
            raise ValueError(
                "The marker length cannot be greater than the square length."
            )

        self._vert_count = vertical_squares_count
        self._horz_count = horizontal_squares_count
        self._square_len = square_len
        self._marker_len = marker_len
        self._resolution = (
            resolution
            if isinstance(resolution, Resolution)
            else Resolution.from_serialized(resolution)
        )

        self._predefined_aruco_dict = aruco_dict
        self._dictionary = cv2.aruco.getPredefinedDictionary(
            self._predefined_aruco_dict
        )

        aruco_indices = list(
            range(0, horizontal_squares_count * vertical_squares_count)
        )
        self._random_seed = seed
        rng = random.Random(seed)
        rng.shuffle(aruco_indices)
        self._dictionary.bytesList = self._dictionary.bytesList[aruco_indices]

        self._board = cv2.aruco.CharucoBoard(
            (self._horz_count, self._vert_count),
            self._square_len,
            self._marker_len,
            self._dictionary,
        )

        if self._horz_count >= self._vert_count:
            board_len = int(
                self._resolution.value * self._square_len * self._horz_count
            )
            img_len = (board_len, int(board_len * self.square_ratio))
        else:
            board_len = int(
                self._resolution.value * self._square_len * self._vert_count
            )
            img_len = (int(board_len / self.square_ratio), board_len)
        self._board_image: np.ndarray = cv2.aruco.CharucoBoard.generateImage(
            self._board, outSize=img_len, marginSize=0
        )

    @property
    def squares_count(self) -> int:
        """
        Get the total number of squares in the Charuco board.
        """
        return self._horz_count * self._vert_count

    @property
    def horizontal_squares_count(self) -> int:
        """
        Get the number of squares in the horizontal direction.
        """
        return self._horz_count

    @property
    def vertical_squares_count(self) -> int:
        """
        Get the number of squares in the vertical direction.
        """
        return self._vert_count

    @property
    def width_len(self) -> float:
        """
        Get the total width of the Charuco board in meters.
        """
        return self._horz_count * self._square_len

    @property
    def height_len(self) -> float:
        """
        Get the total height of the Charuco board in meters.
        """
        return self._vert_count * self._square_len

    @property
    def square_ratio(self) -> float:
        """
        Get the ratio of vertical to horizontal squares.
        """
        return self._vert_count / self._horz_count

    @property
    def aruco_ids(self) -> list[int]:
        """
        Get the list of ArUco marker IDs used in the Charuco board.
        """
        return self._board.getIds().flatten().tolist()

    @property
    def random_seed(self) -> int | None:
        """
        Get the random seed used for shuffling the ArUco markers.
        """
        return self._random_seed

    @property
    def cv2_board_image(self) -> np.ndarray:
        """
        Get the generated Charuco board image as a cv2 image.
        """
        return self._board_image

    def show(self) -> None:
        """
        Show the generated Charuco board image in a window.
        """
        cv2.imshow("img", self._board_image)
        cv2.waitKey()

    @classmethod
    def load(cls, load_folder: Path) -> "Charuco":
        """
        Load a Charuco board from a folder containing the board parameters and image.

        Parameters:
            load_folder (Path): Path to the folder containing the board parameters and image.

        Returns:
            Charuco: Loaded Charuco board object.
        """
        param_path = load_folder / "board.json"
        params = json.load(open(param_path, "r"))
        return cls(**params)

    def save(self, save_folder: Path, override: bool = False) -> None:
        """
        Save the Charuco board parameters and image to a folder.

        Parameters:
            save_folder (Path): Path to the folder where the board parameters and image will be saved
            override (bool): Whether to override the folder if it already exists.
        """
        if not override and save_folder.exists():
            raise FileExistsError(
                f"The folder {save_folder} already exists. Use override=True to overwrite."
            )
        save_folder.mkdir(parents=True, exist_ok=True)

        json.dump(
            self.serialize(), open(save_folder / "board.json", "w"), indent=2
        )
        cv2.imwrite(save_folder / "board.png", self._board_image)

    def serialize(self) -> dict:
        """
        Serialize the Charuco board parameters to a dictionary.

        Returns:
            dict: Dictionary containing the Charuco board parameters.
        """
        return {
            "vertical_squares_count": self._vert_count,
            "horizontal_squares_count": self._horz_count,
            "square_len": self._square_len,
            "marker_len": self._marker_len,
            "resolution": self._resolution.serialized,
            "aruco_dict": self._predefined_aruco_dict,
            "seed": self._random_seed,
        }

    def detect(
        self,
        frame: Frame,
        camera: Camera,
        translation_initial_guess: TranslationVector = None,
        rotation_initial_guess: RotationMatrix = None,
    ) -> tuple[TranslationVector | None, RotationMatrix | None]:
        """
        Detect markers and ChArUco corners in the given image.

        Parameters:
            frame (Frame): Frame to detect markers and corners from.
            camera (Camera): Camera parameters for pose estimation.
            translation_initial_guess (TranslationVector): Initial guess
              for translation vector.
            rotation_initial_guess (RotationMatrix): Initial guess
              for rotation matrix.
        Returns:
            tuple[TranslationVector | None, RotationMatrix | None]: translation vector and rotation matrix.
        """
        result = _detect_marker_corners(frame, self)
        if result is None:
            return None, None
        _, _, charuco_corners, charuco_ids = result

        # Default values
        translation_initial_guess = (
            translation_initial_guess.as_array()
            if translation_initial_guess is not None
            else np.zeros((3, 1))
        )
        rotation_initial_guess = (
            rotation_initial_guess.to_rodrigues()
            if rotation_initial_guess is not None
            else np.zeros((3, 1))
        )

        ret, rotation, translation = cv2.aruco.estimatePoseCharucoBoard(
            charuco_corners,
            charuco_ids,
            self._board,
            camera.matrix,
            camera.distorsion_coefficients,
            rotation_initial_guess,
            translation_initial_guess,
        )
        return (
            TranslationVector.from_array(translation) if ret > 0 else None,
            RotationMatrix(cv2.Rodrigues(rotation)[0]) if ret > 0 else None,
        )

    def draw_aruco_markers(self, frame: Frame) -> None:
        result = _detect_marker_corners(frame, self)
        if result is None:
            return
        corners, ids, charuco_corners, charuco_ids = result

        # Draw the detected markers and corners of the corresponding Charuco board
        cv2.aruco.drawDetectedMarkers(frame.get(), corners, ids)
        _draw_detected_corners_charuco_own(
            frame.get(), charuco_corners, charuco_ids
        )

    def draw_estimated_pose_axes(
        self,
        frame: Frame,
        camera: Camera,
        translation: TranslationVector,
        rotation: RotationMatrix,
        axes_length: float = 1.0,
    ) -> None:
        """
        Draw the axes of reference corresponding to the given pose estimation.

        Parameters:
            frame (Frame): Frame to draw the axes on.
            camera (Camera): Camera parameters for pose estimation.
            translation (TranslationVector): Translation vector of the pose estimation.
            rotation (RotationMatrix): Rotation matrix of the pose estimation.
            axes_length (float): Length in meter of the axes to draw in the output frame.
        """
        cv2.drawFrameAxes(
            frame.get(),
            camera.matrix,
            camera.distorsion_coefficients,
            rotation.to_rodrigues(),
            translation.as_array(),
            axes_length,
        )


def _detect_marker_corners(
    frame: Frame, charuco: Charuco
) -> tuple[list[np.ndarray], np.ndarray, np.ndarray, np.ndarray] | None:
    """
    Detect markers and ChArUco corners in the given image.
    """
    grayscale_frame = frame.get(grayscale=True)
    corners, ids = _detect_markers(grayscale_frame, charuco)
    if not corners or ids is None:
        return None

    # Read chessboard corners between markers
    corner_counts, charuco_corners, charuco_ids = (
        cv2.aruco.interpolateCornersCharuco(
            corners, ids, grayscale_frame, charuco._board
        )
    )
    if corner_counts == 0:
        return None

    return corners, ids, charuco_corners, charuco_ids


def _detect_markers(
    frame: np.ndarray, charuco: Charuco
) -> tuple[list[np.ndarray], np.ndarray | None]:
    """
    Detect ArUco markers in the given image, filtering to only include those belonging to the specified Charuco board.
    """
    corners, ids, _ = cv2.aruco.detectMarkers(frame, charuco._dictionary)
    if not corners or ids is None:
        return [], None

    # Filter out the markers to only include those that belong to the current Charuco board
    # Strategy :
    #   1 - Remove all the markers which are not in the current board
    #   2 - Keep only unique markers IDs and compute the mid point of that reduced board
    #   3 - Reinsert all the markers which appear multiple times (other boards on screen) and
    #       remove them based on nearest distance to the mid-board position
    aruco_ids = charuco.aruco_ids
    corners_tp = []
    ids_tp = []
    for corner, id in zip(corners, ids):
        if id in aruco_ids:
            corners_tp.append(corner)
            ids_tp.append(id)
    ids = np.array(ids_tp)
    corners = corners_tp

    if len(ids) == 0:
        return [], None
    non_unique_ids: list[int] = ids[:, 0].tolist()
    unique_ids: set[int] = set(ids[:, 0].tolist())
    ids_tp = []
    corners_tp = []
    for id in unique_ids:
        if non_unique_ids.count(id) == 1:
            index = non_unique_ids.index(id)
            ids_tp.append(ids[index])
            corners_tp.append(corners[index])
    mid_point = np.mean(
        [np.mean(corner[:, 0, :], axis=0) for corner in corners_tp], axis=0
    )

    for id in unique_ids:
        if non_unique_ids.count(id) == 1:
            # Already processed
            continue

        min_distance = float("inf")
        best_so_far = -1
        all_indices = [i for i, cid in enumerate(ids) if cid[0] == id]
        for i in all_indices:
            corner_center = np.mean(corners[i].squeeze(), axis=0)
            distance = np.linalg.norm(corner_center - mid_point)
            if distance < min_distance:
                min_distance = distance
                best_so_far = i
        ids_tp.append(ids[best_so_far])
        corners_tp.append(corners[best_so_far])
    ids = np.array(ids_tp)
    corners = corners_tp

    return corners, ids


def _draw_detected_corners_charuco_own(img, corners, ids):
    """
    Draw rectangles and IDs to the corners
    """

    rect_size = 5
    id_font = cv2.FONT_HERSHEY_SIMPLEX
    id_scale = 0.5
    id_color = (255, 255, 0)
    rect_thickness = 1

    # Draw rectangels and IDs
    for corner, id in zip(corners, ids):
        corner_x = int(corner[0][0])
        corner_y = int(corner[0][1])
        id_text = "Id: {}".format(str(id[0]))
        id_coord = (corner_x + 2 * rect_size, corner_y + 2 * rect_size)
        cv2.rectangle(
            img,
            (corner_x - rect_size, corner_y - rect_size),
            (corner_x + rect_size, corner_y + rect_size),
            id_color,
            thickness=rect_thickness,
        )
        cv2.putText(img, id_text, id_coord, id_font, id_scale, id_color)

File: charuconstruction_flutter/lib/models/charucos/charuco.dart
import 'dart:io';

import 'package:charuconstruction_flutter/models/charucos/charuco_resolution.dart';
import 'package:charuconstruction_flutter/models/charucos/frame.dart';
import 'package:charuconstruction_flutter/utils/math.dart';
import 'package:charuconstruction_flutter/utils/misc.dart';
import 'package:dartcv4/dartcv.dart';
import 'package:path_provider/path_provider.dart';

class Charuco {
  ///
  /// The total number of squares in the charuco board.
  ///
  int get totalSquaresCount => verticalSquaresCount * horizontalSquaresCount;

  ///
  /// The number of squares in the vertical direction of the charuco board.
  ///
  final int verticalSquaresCount;

  ///
  /// The number of squares in the horizontal direction of the charuco board.
  ///
  final int horizontalSquaresCount;

  ///
  /// The length of the squares in the charuco board, in meters.
  ///
  final double squareLen;

  ///
  /// The length of the markers in the charuco board, in meters.
  /// The markers are the black squares in the charuco board.
  ///
  final double markerLen;

  ///
  /// Get the total width of the Charuco board in meters.
  ///
  double get widthLen => horizontalSquaresCount * squareLen;

  ///
  /// Get the total height of the Charuco board in meters.
  ///
  double get heightLen => verticalSquaresCount * squareLen;

  ///
  /// Get the ratio of vertical to horizontal squares.
  ///
  double get squareRatio => verticalSquaresCount / horizontalSquaresCount;

  ///
  /// The Aruco dictionary used for the markers pattern in the charuco board.
  ///
  final int arucoDictId;
  final ArucoDictionary arucoDict;

  ///
  /// The resolution of the charuco board, in dots per inch (DPI), i.e. the number
  /// of pixels in one inch on the printed page.
  ///
  final CharucoResolution resolution;

  ///
  /// Get the list of ArUco marker IDs used in the Charuco board.
  ///
  List<int> get markerIds => board.ids.toList();

  ///
  /// OpenCV Charuco board generated from the dictionary and the parameters of the charuco board.
  ///
  late final CharucoBoard board;
  late final Mat boardImage;

  ///
  /// Seed for random number generator (will determine the order of aruco markers).
  ///
  final int? seed;

  Charuco({
    required this.verticalSquaresCount,
    required this.horizontalSquaresCount,
    required this.squareLen,
    required this.markerLen,
    required this.resolution,
    required this.arucoDictId,
    this.seed,
  }) : arucoDict = ArucoDictionary.predefined(
         PredefinedDictionaryType.values[arucoDictId],
       ) {
    // Sanity checks
    if (markerLen > squareLen) {
      throw ArgumentError(
        "The marker length cannot be greater than the square length.",
      );
    }

    // Reorder the Aruco markers in the dictionary according to the seed,
    // so that the same seed will always produce the same charuco board.
    _reorderArucoDictionary(
      arucoDict: arucoDict,
      squareCount: horizontalSquaresCount * verticalSquaresCount,
      seed: seed,
    );

    board = CharucoBoard.create(
      (horizontalSquaresCount, verticalSquaresCount),
      squareLen,
      markerLen,
      arucoDict,
    );

    final isHorz = horizontalSquaresCount >= verticalSquaresCount;
    final boardLen =
        resolution.pixelsPerMeter *
        squareLen *
        (isHorz ? horizontalSquaresCount : verticalSquaresCount);
    final imgLen = isHorz
        ? (boardLen, boardLen * squareRatio)
        : (boardLen / squareRatio, boardLen);
    final imgLenAsInt = (imgLen.$1.toInt(), imgLen.$2.toInt());
    boardImage = board.generateImage(imgLenAsInt, marginSize: 0);
  }

  Map<String, dynamic> serialize() {
    return {
      "vertical_squares_count": verticalSquaresCount,
      "horizontal_squares_count": horizontalSquaresCount,
      "square_len": squareLen,
      "marker_len": markerLen,
      "resolution": resolution.name,
      "aruco_dict": arucoDictId,
      "seed": seed,
    };
  }

  static Charuco fromSerialized(Map<String, dynamic> json) {
    return Charuco(
      verticalSquaresCount: json['vertical_squares_count'] as int,
      horizontalSquaresCount: json['horizontal_squares_count'] as int,
      squareLen: json['square_len'] as double,
      markerLen: json['marker_len'] as double,
      resolution: CharucoResolution.fromName(json['resolution'] as String),
      arucoDictId: json['aruco_dict'] as int,
      seed: json['seed'] as int,
    );
  }

  ///
  /// Show the generated Charuco board image in a window.
  ///
  void show() {
    imshow("img", boardImage);
    waitKey(0);
  }

  ///
  /// Save the Charuco board parameters and image to a folder.
  /// [saveFolder] Path to the folder where the board parameters and image will be saved
  /// [override] Whether to override the folder if it already exists.
  ///
  Future<void> save(String saveFolder, {bool override = false}) async {
    // Prepare the save directory
    final baseDir = await getApplicationDocumentsDirectory();
    final saveDir = Directory('${baseDir.path}/$saveFolder');
    if (!override && saveDir.existsSync()) {
      throw FileSystemException(
        "The folder $saveFolder already exists. Use override=true to overwrite.",
      );
    }
    saveDir.createSync(recursive: true);

    // Save the board parameters to a JSON file
    final jsonFile = File('${saveDir.path}/board.json');
    jsonFile.writeAsStringSync(serialize().toString(), flush: true);

    // Save the board image to a PNG file
    imwrite('${saveDir.path}/board.png', boardImage);
  }

  ///
  /// Detect markers and ChArUco corners in the given image.
  /// [frame] OpenCV frame to detect markers and corners from.
  /// [camera] Camera parameters for pose estimation.
  /// [translationInitialGuess] Initial guess for translation vector.
  /// [rotationInitialGuess] Initial guess for rotation matrix.
  void detect({
    required Frame frame,
    // required Camera camera,
    // required TranslationVector? translationInitialGuess,
    // required RotationMatrix? rotationInitialGuess,
  }) {
    final (charucoCorners, charucoIds) =
        _detectMarkerCorners(frame: frame) ?? ([], []);
    if (charucoCorners.isEmpty || charucoIds.isEmpty) return;

    // TODO: Implement initial guesses
    // Initial guesses
    // translation_initial_guess = (
    //     translation_initial_guess.as_array()
    //     if translation_initial_guess is not None
    //     else np.zeros((3, 1))
    // )
    // rotation_initial_guess = (
    //     rotation_initial_guess.to_rodrigues()
    //     if rotation_initial_guess is not None
    //     else np.zeros((3, 1))
    // )

    // was_found, rotation, translation = cv2.aruco.estimatePoseCharucoBoard(
    //     charucoCorners,
    //     charucoIds,
    //     board,
    //     camera.matrix,
    //     camera.distorsion_coefficients,
    //     rotation_initial_guess,
    //     translation_initial_guess,
    // )
    // if not was_found:
    //     return None, None

    // # Reproject the corners to compute the reprojection error and filter out bad detections
    // reprojected_corners, _ = cv2.projectPoints(
    //     self._board.getChessboardCorners()[charuco_ids.flatten()],
    //     rvec=rotation,
    //     tvec=translation,
    //     cameraMatrix=camera.matrix,
    //     distCoeffs=camera.distorsion_coefficients,
    // )
    // error = np.mean(
    //     np.linalg.norm(reprojected_corners - charuco_corners, axis=2)
    // )
    // if error > 5.0:
    //     return None, None

    // # If we get here, the charuco is properly recognized
    // return (
    //     TranslationVector.from_array(translation),
    //     RotationMatrix(cv2.Rodrigues(rotation)[0]),
    // )
  }

  ///
  /// Detect markers and ChArUco corners in the given image.
  ///
  (List<List<Point2f>>, List<int>)? _detectMarkerCorners({
    required Frame frame,
  }) {
    final grayscaleFrame = frame.get(grayscale: true);
    final (corners, ids) = _detectMarkers(frame: grayscaleFrame) ?? ([], []);
    if (corners.isEmpty || ids.isEmpty) return null;

    // // Detect the charuco board
    // final detectorParameters = CharucoDetectorParameters.empty();
    // final detector = CharucoDetector.create(board.board, detectorParameters);
    // final (cornersTp, idsTp, _, _) = detector.detectBoard(frame);

    // corner_counts, charuco_corners, charuco_ids = (
    //     cv2.aruco.interpolateCornersCharuco(
    //         corners, ids, grayscale_frame, charuco._board
    //     )
    // )
    // if corner_counts == 0:
    //     return None

    return null; // TODO return the proper corners and ids
  }

  ///
  /// Detect ArUco markers in the given image, filtering to only include those belonging to the specified Charuco board.
  ///
  (List<Point2f>, List<int>)? _detectMarkers({required InputArray frame}) {
    final detector = ArucoDetector.create(
      arucoDict,
      ArucoDetectorParameters.empty(),
    );
    final (cornersTp, idsTp, _) = detector.detectMarkers(frame);
    if (cornersTp.isEmpty || idsTp.isEmpty) return null;
    if (cornersTp.length != idsTp.length) {
      throw StateError(
        "The number of detected corners and IDs should be the same.",
      );
    }

    //
    // Filter out the markers to only include those that belong to the current Charuco board
    // Strategy :
    //   1 - Remove all the markers which are not in the current board
    //   2 - Keep only unique markers IDs and compute the mid point of that reduced board
    //   3 - Reinsert all the markers which appear multiple times (other boards on screen) and
    //       remove them based on nearest distance to the mid-board position

    // Step 1 - Find all the markers that belong to the current board. It may include
    // markers shared with other boards at that point.
    final ids = <int>[];
    final corners = <Point2f>[];
    for (int i = 0; i < idsTp.length; i++) {
      if (markerIds.contains(idsTp[i])) {
        ids.add(idsTp[i]);
        corners.addAll(cornersTp[i].toList());
      }
    }
    if (ids.isEmpty) return null;

    // Step 2 - Keep only the unique IDs, which are definitely from the current board
    // to compute the mid point of the board.
    final uniqueIndices = ids.singleItemsIndices();
    final pointCount = uniqueIndices.length.toDouble();
    final midPoint = uniqueIndices
        .map((i) => corners[i])
        .fold(
          Point2f(0, 0),
          (weightedSum, corner) => Point2f(
            weightedSum.x + (corner.x / pointCount),
            weightedSum.y + (corner.y / pointCount),
          ),
        );

    // Step 3 - Insert all the markers. For those which appears multiple times
    // (meaning, markers shared across multiple boards), we need to determine which
    // one belongs to the current board by keeping only the one with the nearest
    // distance to the mid-board position
    final idsOfCurrentBoard = <int>[];
    final cornersOfCurrentBoard = <Point2f>[];
    for (int i = 0; i < ids.length; i++) {
      if (idsOfCurrentBoard.contains(ids[i])) {
        // This marker has already been processed, we skip it
        continue;
      }

      if (uniqueIndices.contains(i)) {
        // This is a unique marker, we keep it
        idsOfCurrentBoard.add(ids[i]);
        cornersOfCurrentBoard.add(corners[i]);
        continue;
      }

      // Find all the indices of that marker ID
      final allIndices = [
        for (int j = 0; j < ids.length; j++)
          if (ids[j] == ids[i]) j,
      ];

      var minDistance = double.infinity;
      var bestSoFar = -1;
      for (final index in allIndices) {
        final corner = corners[index];
        final x = corner.x - midPoint.x;
        final y = corner.y - midPoint.y;
        final distanceSquared = x * x + y * y;
        if (distanceSquared < minDistance) {
          minDistance = distanceSquared;
          bestSoFar = index;
        }
      }
      if (i != bestSoFar) {
        // This marker appears multiple times, but the one at index "i" is not the
        // closest to the mid-board position, so we skip it. This is important
        // to preserve the order of the markers
        continue;
      }
      idsOfCurrentBoard.add(ids[bestSoFar]);
      cornersOfCurrentBoard.add(corners[bestSoFar]);
    }

    return (cornersOfCurrentBoard, idsOfCurrentBoard);
  }
}

void _reorderArucoDictionary({
  required ArucoDictionary arucoDict,
  required int squareCount,
  int? seed,
}) {
  final random = PythonRandom(seed: seed);
  final arucoIndices = List.generate(squareCount, (i) => i);
  random.shuffle(arucoIndices);

  final original = arucoDict.bytesList;
  final reordered = Mat.zeros(squareCount, original.cols, original.type);
  for (int i = 0; i < squareCount; i++) {
    original.row(arucoIndices[i]).copyTo(reordered.row(i));
  }

  arucoDict.bytesList = reordered;
}
